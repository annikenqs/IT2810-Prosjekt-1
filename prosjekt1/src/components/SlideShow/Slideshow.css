.slideshow-container {
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	text-align: center;
	margin: 0 auto;
}

.jokeIDButton-container {
	display: grid;
	grid-template-columns: repeat(auto-fit, minmax(45px, 1fr));
	grid-gap: 10px;
	margin-bottom: 30px;
	margin-left: auto;
	margin-right: auto;
	padding: 10px;
	max-width: 80vw;
}

.jokeIDButton {
	background-color: #8a95ba;
	color: #fff;
	border: none;
	border-radius: 8px;
	font-size: 16px;
	width: 45px;
	height: 45px;
	cursor: pointer;
	transition: background-color 0.3s ease;
}

.jokeIDButton:hover {
	background-color: #c9e2dd;
	color: #333;
}

@media (max-width: 600px) {
	.jokeIDButton {
		flex: 1 1 9%;
	}
}

.skipToJoke {
	color: #fff;
	text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.2);
}

.card-wrapper {
  position: relative;
  display: flex;
  justify-content: center;
  width: 75%;
  max-width: 500px;
  margin: auto;
}

.arrow {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  background: none;
  border: none;
  font-size: 4rem;
  font-weight: bold;
  color: white;
  cursor: pointer;
  transition: color 0.3s ease;
  user-select: none;
  z-index: 1;
}

.arrow:hover {
  color: #c9e2dd;
}

.left-arrow {
  left: 10px;
}

.right-arrow {
  right: 10px;
}

@media (max-width: 600px) {
  .arrow {
    font-size: 2.5rem;
  }

  .left-arrow {
    left: 0;
  }

  .right-arrow {
    right: 0;
  }

  .card-wrapper {
    padding: 0 30px;
  }

  .slideshow-container {
    padding: 5px;
  }
}