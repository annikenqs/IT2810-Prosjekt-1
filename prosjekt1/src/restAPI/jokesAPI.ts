import { useQuery} from '@tanstack/react-query';

export interface JokeResponse {
	type: string;
	joke?: string; // For single-type jokes
	setup?: string; // For two-part jokes
	delivery?: string; // For two-part jokes
	id: number;
	category: string;
}

export const validIDs: number[] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 241, 243, 244, 245, 249, 250, 251, 252, 253, 293, 183, 295, 296, 297, 298, 299, 300, 311, 313, 315];

// Fetch a joke, given a category
const fetchJoke = async (type: string): Promise<JokeResponse> => {
	const response = await fetch(`https://v2.jokeapi.dev/${type}?blacklistFlags=racist,sexist&idRange=0-318`);
	if (!response.ok) {
		throw new Error("Failed to fetch joke");
	} else if (!(type=='Christmas' || type=='Spooky' || type=='Programming')) {
		throw new Error("Invalid joke category");
	}
	const data: JokeResponse = await response.json();
	return data;
}

//Fetch a joke, given an ID (the ID has to be valid)
const fetchJokeById = async (id: number): Promise<JokeResponse> => {
	if (!validIDs.includes(id)) {
		throw new Error("Invalid joke ID");
	}
	const response = await fetch(`https://v2.jokeapi.dev/joke/Programming,Spooky,Christmas?blacklistFlags=racist,sexist&idRange=${id}`);
	if (!response.ok) {
		throw new Error("Failed to fetch joke");
	}
	const data: JokeResponse = await response.json();
	return data;
}

//Fetch jokes for a specific category
const fetchJokesByCategory = async (category: string, batchSize: number, rangeStart: number, rangeEnd: number): Promise<JokeResponse[]> => {
  const response = await fetch(
    `https://v2.jokeapi.dev/joke/${category}?blacklistFlags=racist,sexist&amount=${batchSize}&idRange=${rangeStart}-${rangeEnd}`
  );
  if (!response.ok) {
    throw new Error(`Failed to fetch ${category} jokes`);
  }
  const data = await response.json();
  if (data.error || (data.jokes && data.jokes.length === 0)) {
    console.log(`No jokes found for ${category}. Error:`, data.error ? data.error : 'None found');
    return [];
  }
  return data.jokes ?? [];
};

// Fetch exactly 10 jokes from each category
const fetchAllJokes = async (): Promise<JokeResponse[]> => {
  const batchSize = 10;
  let allJokes: JokeResponse[] = [];
	//Christmas IDs: 241, 243 - 245, 249, 250 - 253, 293
  const christmasJokes = await fetchJokesByCategory("Christmas", batchSize, 241, 293);
  if (christmasJokes.length == 10) allJokes = [...allJokes, ...christmasJokes];
	//Spooky IDs: 183, 295 - 300, 311, 313, 315
  const spookyJokes: JokeResponse[] = await fetchJokesByCategory("Spooky", batchSize, 183, 315);
  if (spookyJokes.length == 10) allJokes = [...allJokes, ...spookyJokes];
	//Programming IDs: 0-9
  const programmingJokes = await fetchJokesByCategory("Programming", batchSize, 0, 10);
  if (programmingJokes.length == 10) allJokes = [...allJokes, ...programmingJokes];
  return allJokes;
};

// Get a random joke
export const useRandomJoke = () => {
	return useQuery({
		queryKey:["joke", "Any"],
		 queryFn: () => fetchJoke("Any"),
	})
}

// Get a programming joke
export const useProgrammingJoke = () => {

	return useQuery({
		queryKey: ["joke", "Programming"],
		queryFn: () => fetchJoke("Programming"),
	});
};

// Get a spooky joke
export const useSpookyJoke = () => {

	return useQuery({
		queryKey: ["joke", "Spooky"],
		queryFn: () => fetchJoke("Spooky"),
	});
};

// Get a christmas joke
export const useChristmasJoke = () => {
	
	return useQuery({
		queryKey: ["joke", "Christmas"],
		queryFn: () => fetchJoke("Christmas"),
	});
};


// Get a joke by ID
export const useJokeById = (id: number) => {
	return useQuery({
		queryKey: ["joke", id],
		queryFn: () => fetchJokeById(id),
		refetchInterval: 1000 * 60 * 10, // Refetch every 10 min
	});
};

//Get all relevant jokes
export const useAllJokes = () => {
  return useQuery({
    queryKey: ["joke", "All"],
    queryFn: () => fetchAllJokes(),
		refetchInterval: 1000 * 60 * 10,
  });
};