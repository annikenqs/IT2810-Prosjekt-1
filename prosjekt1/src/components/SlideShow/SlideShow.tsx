import { useState } from "react";
import "../JokeCard/JokeCard";
import "./SlideShow.css";
import "../../index.css";
import JokeCard from "../JokeCard/JokeCard";
import {validIDs} from "../../restAPI/jokesAPI"; 


function JokeSlideshow() {
  const [currentIndex, setCurrentIndex] = useState<number>(0);

  // Hent vitsen basert pÃ¥ gjeldende ID fra validIDs (ved bruk av index i listen)
  const currentJokeId = validIDs[currentIndex];


  // Funksjon for neste vits
  const nextJoke = () => {
    setCurrentIndex((prevIndex) => 
      prevIndex < validIDs.length - 1 ? prevIndex + 1 : 0 
    );
  };

  // Funksjon for forrige vits
  const previousJoke = () => {
    setCurrentIndex((prevIndex) => 
      prevIndex > 0 ? prevIndex - 1 : validIDs.length - 1 
    );
  };


  return (
    <div className="slideshow-container">
        <JokeCard jokeId={currentJokeId}/>
      <div className="controls">
        <button className="button" onClick={previousJoke}>
          Previous
        </button>
        <button className="button" onClick={nextJoke}>
          Next
        </button>
      </div>
    </div>
  );
}

export default JokeSlideshow;
