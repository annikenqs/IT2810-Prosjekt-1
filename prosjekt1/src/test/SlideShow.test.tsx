import React from "react";
import { describe, it, beforeEach, afterEach, vi } from "vitest";
import { cleanup, fireEvent, render, screen } from "@testing-library/react";
import SlideShow from "../components/SlideShow/SlideShow";
import { useJokeById, slideshowIDs } from "../restAPI/jokesAPI";

vi.mock("../restAPI/jokesAPI", async (importOriginal) => {
	const actual = (await importOriginal()) as Record<string, unknown>;
	return {
		...actual,
		slideshowIDs: [0, 1, 2],
		inUseIDs: [0, 1, 2, 3, 4],
		useJokeById: vi.fn(),
	};
});

describe("SlideShow", () => {
	const mockJokeResponse = {
		id: 0,
		type: "single",
		joke: "I've got a really good UDP joke to tell you but I don’t know if you'll get it.",
		isLoading: false,
		error: null,
		category: "Programming",
	};

	beforeEach(() => {
		// Specify type for the mock return value
		(useJokeById as jest.Mock).mockReturnValue({
			data: mockJokeResponse,
		});
	});

	afterEach(() => {
		cleanup();
	});

	it("renders SlideShow correctly and matches snapshot", () => {
		const { container } = render(<SlideShow />);
		expect(container).toMatchSnapshot();
	});

	it("displays current joke and allows navigating to next and previous jokes", async () => {
		let currentIndex = 0;
		
		// Mock the behavior of useJokeById to update based on currentIndex
		(useJokeById as jest.Mock).mockImplementation(() => ({
			data: { ...mockJokeResponse, id: slideshowIDs[currentIndex] },
		}));
		render(<SlideShow />);

		// Expect the first joke to be displayed
		expect(
			screen.getByText(
				/I've got a really good UDP joke to tell you but I don’t know if you'll get it./,
			),
		).toBeInTheDocument();

		// Click Next and expect the next joke to load
		currentIndex = 1;
		fireEvent.click(screen.getByRole("button", { name: /›/ }));
		expect(useJokeById).toHaveBeenCalledWith(slideshowIDs[1]);

		// Click Previous and expect the first joke to load again
		currentIndex = 0;
		fireEvent.click(screen.getByRole("button", { name: /‹/ }));
		expect(useJokeById).toHaveBeenCalledWith(slideshowIDs[0]);
	});

	it("allows skipping to a specific joke by joke ID", () => {
		render(<SlideShow />);

		// Click on the joke ID button for the second joke
		fireEvent.click(screen.getByText("2"));
		expect(useJokeById).toHaveBeenCalledWith(slideshowIDs[1]);

		// Click on the joke ID button for the first joke
		fireEvent.click(screen.getByText("1"));
		expect(useJokeById).toHaveBeenCalledWith(slideshowIDs[0]);
	});

	it("displays loading message when joke is loading", () => {
		// Mock loading state
		(useJokeById as jest.Mock).mockReturnValue({
			data: null,
		});

		render(<SlideShow />);

		expect(screen.getByText(/Loading/)).toBeInTheDocument();
	});

	it("displays error message when there's an error", () => {
		// Mock error state
		(useJokeById as jest.Mock).mockReturnValue({
			data: { ...mockJokeResponse, error: { message: "Failed to load joke" } },
		});

		render(<SlideShow />);

		expect(screen.getByText(/Failed to load joke/)).toBeInTheDocument();
	});

	it("rolls over to the first joke when Next is clicked on the last joke", () => {
		render(<SlideShow />);

		// Mock `useJokeById` to handle current joke as the last joke in slideshowIDs
		(useJokeById as jest.Mock).mockReturnValue({
			data: { ...mockJokeResponse, id: slideshowIDs[slideshowIDs.length - 1] },
		});

		// Simulate clicking the "Next" button on the last joke
		fireEvent.click(screen.getByRole("button", { name: /›/ }));

		// Expect `useJokeById` to be called with the first joke ID (rollover behavior)
		expect(useJokeById).toHaveBeenCalledWith(slideshowIDs[0]);
	});

	it("rolls over to the last joke when Previous is clicked on the first joke", () => {
		render(<SlideShow />);

		// Mock `useJokeById` to handle current joke as the first joke in slideshowIDs
		(useJokeById as jest.Mock).mockReturnValue({
			data: { ...mockJokeResponse, id: slideshowIDs[0] },
		});

		// Simulate clicking the "Previous" button on the first joke
		fireEvent.click(screen.getByRole("button", { name: /‹/ }));

		// Expect `useJokeById` to be called with the last joke ID (rollover behavior)
		expect(useJokeById).toHaveBeenCalledWith(slideshowIDs[slideshowIDs.length - 1]);
	});

	it("navigates between jokes and handles rollovers", () => {
		render(<SlideShow />);

		// Simulate clicking the "Next" button and check navigation
		fireEvent.click(screen.getByRole("button", { name: /›/ }));
		expect(useJokeById).toHaveBeenCalledWith(slideshowIDs[1]);

		// Simulate clicking "Next" at the end and check rollover to first joke
		fireEvent.click(screen.getByRole("button", { name: /›/ }));
		fireEvent.click(screen.getByRole("button", { name: /›/ })); // This would be the third click
		expect(useJokeById).toHaveBeenCalledWith(slideshowIDs[0]);

		// Simulate clicking "Previous" and check navigation
		fireEvent.click(screen.getByRole("button", { name: /‹/ }));
		expect(useJokeById).toHaveBeenCalledWith(slideshowIDs[2]);

		// Simulate clicking "Previous" at the start and check rollover to last joke
		fireEvent.click(screen.getByRole("button", { name: /‹/ }));
		fireEvent.click(screen.getByRole("button", { name: /‹/ }));
		expect(useJokeById).toHaveBeenCalledWith(slideshowIDs[slideshowIDs.length - 1]);
	});
});
