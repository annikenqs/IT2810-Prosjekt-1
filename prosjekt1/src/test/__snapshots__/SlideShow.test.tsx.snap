// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SlideShow > renders SlideShow correctly and matches snapshot 1`] = `
<div>
  <div
    class="slideshow-container"
  >
    <section
      class="card"
      role="figure"
    >
      <h3>
        Joke #
        1
      </h3>
      <p>
        - 
        I've got a really good UDP joke to tell you but I don’t know if you'll get it.
      </p>
      <button
        class="favorite-button"
      >
        ☆
      </button>
    </section>
    <div
      class="controls"
    >
      <button
        class="button"
      >
        Previous
      </button>
      <button
        class="button"
      >
        Next
      </button>
    </div>
    <h3
      class="skipToJoke"
    >
      Skip to joke
    </h3>
    <div
      class="jokeIDButton-container"
    >
      <button
        class="jokeIDButton"
      >
        1
      </button>
      <button
        class="jokeIDButton"
      >
        2
      </button>
      <button
        class="jokeIDButton"
      >
        3
      </button>
    </div>
  </div>
</div>
`;
